Services/Administration/OJS.Services.Administration.Business/Problems/ProblemsBusinessService.cs

```diff
@@ -235,6 +244,96 @@ public override async Task<ProblemAdministrationModel> Edit(ProblemAdministratio
+            .AsEnumerable()
+            .Select(s => (s.CompletedExecutionOn.Value - s.StartedExecutionOn.Value).TotalSeconds)
+            .Average();
+
+        var maxWorkersWorkingTime =
+            await this.settingsBusinessService.GetByKey(GlobalConstants.Settings.MaxWorkersWorkingTimeInSeconds);
+        var maxSubmissionsCountAllowedForBatchRetest =
+            await this.settingsBusinessService.GetByKey(GlobalConstants.Settings.MaxSubmissionsCountAllowedForBatchRetest);
+        var maxSubmissionTimeToExecuteAllowedForBatchRetest =
+            await this.settingsBusinessService.GetByKey(GlobalConstants.Settings.MaxSubmissionTimeToExecuteAllowedForBatchRetest);
+
+        var allSubmissionsWorkingTime = submissionsCount * averageTimeDifferenceInSeconds;
+
+        var allSubmissionsWorkingTimeExceedsMaxAllowedTime = allSubmissionsWorkingTime > double.Parse(maxWorkersWorkingTime.Value, CultureInfo.InvariantCulture);
+        var submissionsCountExceedsMaxAllowedLimit = submissionsCount > int.Parse(maxSubmissionsCountAllowedForBatchRetest.Value!, CultureInfo.InvariantCulture);
+        var submissionsTimeToExecuteExceedsMaxAllowedLimit = averageTimeDifferenceInSeconds < int.Parse(maxSubmissionTimeToExecuteAllowedForBatchRetest.Value!, CultureInfo.InvariantCulture);
+
+        var validationModel = new ProblemRetestValidationModel
+        {
+            SubmissionsCount = submissionsCount,
+            AverageExecutionTime = averageTimeDifferenceInSeconds,
+            RetestAllowed = false,
+        };
+
+        var canRetest = !allSubmissionsWorkingTimeExceedsMaxAllowedTime || (!submissionsCountExceedsMaxAllowedLimit && !submissionsTimeToExecuteExceedsMaxAllowedLimit);
+
+        if (this.userProviderService.GetCurrentUser().IsDeveloper && !canRetest)
+        {
+            // Developers can retest even if validations fail
+            // Adding messages for developers for better understanding why retest fails
+            if (allSubmissionsWorkingTimeExceedsMaxAllowedTime)
+            {
+                validationModel.Message = $"Submissions will take {Math.Round(allSubmissionsWorkingTime / 60)} minutes to execute.";
+            }
+            else if (submissionsCountExceedsMaxAllowedLimit)
+            {
+                validationModel.Message = $"Submissions count exceeds max allowed limit {maxSubmissionsCountAllowedForBatchRetest.Value}.";
+            }
+            else if (submissionsTimeToExecuteExceedsMaxAllowedLimit)
+            {
+                validationModel.Message =
+                    $"Submissions time to execute ({averageTimeDifferenceInSeconds}) exceeds max allowed limit {maxSubmissionTimeToExecuteAllowedForBatchRetest.Value}";
+            }
+        }
+        else if (!canRetest)
+        {
+            validationModel.Message =
+                $"Unable to retest {submissionsCount} submissions for this problem. Contact a developer for this action.";
```

Comment:
Can't we provide a clearer message here?

Reply:
Fixed